### DISCUSSION TITLE
Random order of blocks in diagrams

### DISCUSSION THREAD

**Discussion by antonostrovsky** [christianvoigt/argdown#255] [Q&A]
First of all, thank you for the great project! I really enjoy using Argdown and have started to use it for documenting Yale Uni lectures: https://github.com/antonostrovsky/open-yale-death-shelly-kagan

One problem that really stops me from creating useful diagrams is that each time I add a new character, the order of blocks on the diagram changes randomly.
Example: https://github.com/antonostrovsky/open-yale-death-shelly-kagan/blob/master/lectures/9.%20Plato%2C%20Part%20IV%2C%20Arguments%20for%20the%20immortality%20of%20the%20soul%20(cont.).ad

Try adding absolutely any character in any place in the document and the parts will randomly re-shuffle in the Viz.JS or Dagre preview.

Is there any way I can enforce the order of the parts (grey background blocks that are created by starting a paragraph with ##)?

Best regards,
Anton

**Comment by christianvoigt** [comment]
Hi Anton,
We are using the layout algorithms of VizJs (graphviz) and Dagre. Both are using a layer or rank-based hierarchical layout (see [here](http://docs.yworks.com/yfiles/doc/developers-guide/incremental_hierarchical_layouter.html) for a short explanation). 

take a look at: https://argdown.org/guide/changing-the-graph-layout.html

At the moment Dagre does not offer us many configuration options, but I think you might find the VizJs `rank`, `samerank` and `newrank` options useful, though there is currently no way to change the "sequence" within the same rank/layer (which is what you were looking for).

If you want to have complete control over your graph layout, [we recommend](https://argdown.org/guide/changing-the-graph-layout.html#layouting-in-the-yed-graph-editor) to export your graph to graphml and edit it in [yEd](https://www.yworks.com/products/yed), a free graph editor with amazing features. 

It would be nice to have the option to manually layout the graph directly in VSCode, so if anyone wants to write a custom graph editor for the Argdown extension we would be very happy about that. But it will probably not happen soon.

**Comment by antonostrovsky** [comment]
To be fair, Viz.JS works incredibly well. This is the grouping that the algorithm offers and I couldn't be happier in this case: 
https://github.com/antonostrovsky/open-yale-death-shelly-kagan/raw/master/lectures/17.%20The%20badness%20of%20death%2C%20Part%20II%2C%20The%20deprivation%20account.pdf

**Comment by antonostrovsky** [comment]
As official Argdown documentation suggests, this question on SO provides a plenty of useful examples of using **rank** property: https://stackoverflow.com/questions/25734244/how-do-i-place-nodes-on-the-same-level-in-dot
